guardian/newsletters-nx
# Dblatcher user permissions format

## What does this change?

We currently define who has permissions to make changes in the tool by writing a parameter that assigns a ‘Permission Level' (viewer, editor, drafter, developer) to people by their email address. For each Permission Level, there was a set of actions a user in the level can take.

This worked fine, but as we add more the fine-tuned permissions needed to send emails asking groups of people to (for example) mark a sign-up page as complete, we need to create more levels. 

This model will make it hard to manage users who need overlapping permissions - “Bob is in Central Production, so needs that Permission Level, but also needs to be able to Edit Drafts because…”

To give us more fine-tuned control, we can remove the concept of "Permission Level" and instead have a set of permissions to do specific things and list of the users that can do those things - this is more long-winded, but allows any user to have any combination of permissions.

This PR models permissions using the same structure as https://github.com/guardian/permissions/ with a view to eventually using the existing Permissions App to set who has what permissions on the newsletters tool.


## How to test

Locally:
Update the format for the test permissions in ".env.local" to match "env.local.example" - the test user(software.developer@guardian.co.uk) should have all permissions again.

Setting `active` to false on any of the  `Overrides` (or removing the `Override`) will take away the permissions.

On CODE:
Would need to convert the current env param in AWS to the new Permissions format, then deploy this branch - if it works, the user permission level will be unaffected.

If we deploy to CODE without updating the env param, no users will have any permissions.

## How can we measure success?

User permissions modelled in a more flexible/sustainable way. 

## Have we considered potential risks?

May become more complex to set the Permissions JSON manually.

Will need to update the PROD permissions JSON before merging!
